### PR TITLE
[Sofa.Core][SofaBaseTopology] Refactor TopologyContainer

### PR DESCRIPTION
**Pull request by fredroy**
First steps to move TopologyData/Change into Sofa.Core (see  #2402)

The goal of this PR is to remove *TopologyContainer dependencies in TopologyData mechanism.
TopologyHanderData has direct dependencies on *TopologyContainer because it wants to link their specific Data(d_edges, d_triangles, etc) into a list of engines handled by itself.
It should not know how a topology is managing its data IMO.
So this PR:
 - is abstracting the function to add a Data to list of engines 
 - change the way how the engine lists is managed: each Subclass was setting its engine list (edgeset and then triangleset and then....), now the base class is storing a vector of list of engines (indexed on the type of topology)
 - removes the direct dependency on TopologyContainer in TopologyHandlerData.inl

After this PR, all Topology*Data mechanism will be moved in to Sofa.Core, along Topology* class (sofa::core::topology)

NB: I let the code building a dep graph which was done for debugging purpose, but it is matter of discussion to keep it or not)
 

______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

**Comment by fredroy** [comment]
[ci-build][with-all-tests]

**Comment by fredroy** [comment]
[ci-build][with-all-tests]

File: SofaKernel/modules/SofaBaseTopology/src/SofaBaseTopology/EdgeSetTopologyContainer.cpp [hunk 1]
```diff
@@ -544,44 +544,42 @@ void EdgeSetTopologyContainer::setEdgeTopologyToDirty()
...
 
 void EdgeSetTopologyContainer::cleanEdgeTopologyFromDirty()
 {
     m_edgeTopologyDirty = false;
 
     // security, clean all engines to avoid loops
-    std::list<sofa::core::topology::TopologyHandler *>::iterator it;
-    for ( it = m_enginesList.begin(); it!=m_enginesList.end(); ++it)
+    auto& edgeTopologyHandlerList = getTopologyHandlerList(sofa::geometry::ElementType::EDGE);
+    for (auto topoEngine : edgeTopologyHandlerList)
```
Comment (epernod) [inline SofaKernel/modules/SofaBaseTopology/src/SofaBaseTopology/EdgeSetTopologyContainer.cpp]: Could be the good timing to rename everywhere : topoEngine by topoHandler

Comment (fredroy) [reply SofaKernel/modules/SofaBaseTopology/src/SofaBaseTopology/EdgeSetTopologyContainer.cpp]: okidoki

**Review by epernod — APPROVED** [review]

**Comment by epernod** [comment]
Wait for 7 days, and good to go.

**Pull request merged**